<?xml version="1.0" encoding="UTF-8"?>

<!-- Checking done.. all the bots spawn in gazebo and rviz, the wheels still clump at the chassis centre-->

<launch>
	<!-- Robot pose -->
	<arg name="x" default="0"/>
	<arg name="y" default="0"/>
	<arg name="z" default="0"/>
	<arg name="roll" default="0"/>
	<arg name="pitch" default="0"/>
	<arg name="yaw" default="0"/>
	

	<!-- Launch Gazebo World -->
	<include file="$(find gazebo_ros)/launch/empty_world.launch">

		<arg name="world_name" value="$(find fredbots)/worlds/all_q_world_min_wall.world"/>
		<arg name="paused" value="false"/>
		<arg name="use_sim_time" value="true"/>
		<arg name="gui" value="true"/>
		<arg name="headless" value="false"/>
		<arg name="debug" value="false"/>
    	</include>


  <!-- <node name="atom_tf_broadcaster_wo" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 world odom" /> -->
  <param name="robot_description" command="$(find xacro)/xacro '$(find fredbots)/src/atom/urdf/atom.xacro'" />


  <!-- BEGIN ROBOT 1-->
  <param name="robot_description_atom1" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_1/atom.xacro'" />

  <group ns="atom_1">

    <param name="tf_prefix" value="atom1_tf" />

    <include file="$(find fredbots)/launch/one_robot.launch" >
      <arg name="init_pose" value="-x -5 -y 1 -z 0" />
      <arg name="robot_name"  value="atom1" />

    </include>

    <node name="atom1_tf_broadcaster_wo" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 /atom_1/odom atom1_tf/robot_footprint" /> <!-- use gazebo ns here-->
    <!-- <node name="atom1_tf_broadcaster_rwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom1_tf/chassis atom1_tf/right_wheel_front" />
    <node name="atom1_tf_broadcaster_rwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom1_tf/chassis atom1_tf/right_wheel_back" />
    <node name="atom1_tf_broadcaster_lwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom1_tf/chassis atom1_tf/left_wheel_front" />
    <node name="atom1_tf_broadcaster_lwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom1_tf/chassis atom1_tf/left_wheel_back" /> -->

    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_atom1">

      <param name="robot_description" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_1/atom.xacro'" />
      <param name="use_gui" value="true"/>

    </node>

    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping_atom1" args="scan:=/scan_1" output="screen">


    <param name="base_frame" value="/atom1_tf/robot_footprint"/>
    <param name="odom_frame" value="/atom1_tf/odom"/>
    <param name="map_frame" value="/atom1_tf/map"/>
    <param name="delta" value="0.1"/>
    <param name="maxUrange" value="10.0"/>
    <param name="sigma" value="0.05"/>
    <param name="kernelSize" value="3"/>
    <param name="lstep" value="0.05"/>
    <param name="astep" value="0.05"/>
    <param name="iterations" value="5"/>
    <param name="lsigma" value="0.075"/>
    <param name="ogain" value="3.0"/>
    <param name="minimumScore" value="100"/>
    <param name="maximumScore" value="0"/>
    <param name="publish_map" value="true"/>	

  
  </node>
    

  </group>

  <!-- BEGIN ROBOT 2 -->
	<param name="robot_description_atom2" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_2/atom.xacro'" />

  <group ns="atom_2">

    <param name="tf_prefix" value="atom2_tf" />
    <include file="$(find fredbots)/launch/one_robot.launch" >
      <arg name="init_pose" value="-x 0 -y 5 -z 0" />
      <arg name="robot_name"  value="atom2" />

    </include>

    <node name="atom2_tf_broadcaster_wo" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 /atom_2/odom atom2_tf/robot_footprint" /> <!-- use gazebo ns here-->
    <!-- <node name="atom2_tf_broadcaster_rwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom2_tf/chassis atom2_tf/right_wheel_front" />
    <node name="atom2_tf_broadcaster_rwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom2_tf/chassis atom2_tf/right_wheel_back" />
    <node name="atom2_tf_broadcaster_lwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom2_tf/chassis atom2_tf/left_wheel_front" />
    <node name="atom2_tf_broadcaster_lwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom2_tf/chassis atom2_tf/left_wheel_back" /> -->

    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_atom2">

      <param name="robot_description" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_2/atom.xacro'" />
      <param name="use_gui" value="true"/>

    </node>

    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping_atom2" args="scan:=/scan_2" output="screen">


    <param name="base_frame" value="/atom2_tf/robot_footprint"/>
    <param name="odom_frame" value="/atom2_tf/odom"/>
    <param name="map_frame" value="/atom2_tf/map"/>
    <param name="delta" value="0.1"/>
    <param name="maxUrange" value="10.0"/>
    <param name="sigma" value="0.05"/>
    <param name="kernelSize" value="3"/>
    <param name="lstep" value="0.05"/>
    <param name="astep" value="0.05"/>
    <param name="iterations" value="5"/>
    <param name="lsigma" value="0.075"/>
    <param name="ogain" value="3.0"/>
    <param name="minimumScore" value="100"/>
    <param name="maximumScore" value="0"/>
    <param name="publish_map" value="true"/>	

  
  </node>
    

  </group>

  <!-- BEGIN ROBOT 3-->
  <param name="robot_description_atom3" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_3/atom.xacro'" />

  <group ns="atom_3">

    <param name="tf_prefix" value="atom3_tf" />
    <include file="$(find fredbots)/launch/one_robot.launch" >
      <arg name="init_pose" value="-x -3 -y 4 -z 0" />
      <arg name="robot_name"  value="atom3" />

    </include>

    <node name="atom3_tf_broadcaster_wo" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 /atom_3/odom atom3_tf/robot_footprint" /> <!-- use gazebo ns here-->
    <!-- <node name="atom3_tf_broadcaster_rwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom3_tf/chassis atom3_tf/right_wheel_front" />
    <node name="atom3_tf_broadcaster_rwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom3_tf/chassis atom3_tf/right_wheel_back" />
    <node name="atom3_tf_broadcaster_lwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom3_tf/chassis atom3_tf/left_wheel_front" />
    <node name="atom3_tf_broadcaster_lwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom3_tf/chassis atom3_tf/left_wheel_back" /> -->

    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_atom3">

      <param name="robot_description" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_3/atom.xacro'" />
      <param name="use_gui" value="true"/>

    </node>

    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping_atom3" args="scan:=/scan_3" output="screen">

  
    <param name="base_frame" value="/atom3_tf/robot_footprint"/>
    <param name="odom_frame" value="/atom3_tf/odom"/>
    <param name="map_frame" value="/atom3_tf/map"/>
    <param name="delta" value="0.1"/>
    <param name="maxUrange" value="10.0"/>
    <param name="sigma" value="0.05"/>
    <param name="kernelSize" value="3"/>
    <param name="lstep" value="0.05"/>
    <param name="astep" value="0.05"/>
    <param name="iterations" value="5"/>
    <param name="lsigma" value="0.075"/>
    <param name="ogain" value="3.0"/>
    <param name="minimumScore" value="100"/>
    <param name="maximumScore" value="0"/>
    <param name="publish_map" value="true"/>	

  
  </node>
    

  </group>

  <!-- BEGIN ROBOT 4-->
  <param name="robot_description_atom4" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_4/atom.xacro'" />

  <group ns="atom_4">

    <param name="tf_prefix" value="atom4_tf" />
    <include file="$(find fredbots)/launch/one_robot.launch" >
      <arg name="init_pose" value="-x -2 -y 7 -z 0" />
      <arg name="robot_name"  value="atom4" />

    </include>

    <node name="atom4_tf_broadcaster_wo" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 /atom_4/odom atom4_tf/robot_footprint" /> <!-- use gazebo ns here-->
    <!-- <node name="atom4_tf_broadcaster_rwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom4_tf/chassis atom4_tf/right_wheel_front" />
    <node name="atom4_tf_broadcaster_rwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom4_tf/chassis atom4_tf/right_wheel_back" />
    <node name="atom4_tf_broadcaster_lwf" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom4_tf/chassis atom4_tf/left_wheel_front" />
    <node name="atom4_tf_broadcaster_lwb" pkg="tf2_ros" type="static_transform_publisher"  args="0 0 0 0 0 0 /atom4_tf/chassis atom4_tf/left_wheel_back" /> -->


   <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_atom4">

      <param name="robot_description" command="$(find xacro)/xacro '$(find fredbots)/src/atom/multi_urdf/bot_4/atom.xacro'" />
      <param name="use_gui" value="true"/>

    </node>


    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping_atom4" args="scan:=/scan_4" output="screen">

    <param name="base_frame" value="/atom4_tf/robot_footprint"/>
    <param name="odom_frame" value="/atom4_tf/odom"/>
    <param name="map_frame" value="/atom4_tf/map"/>
    <param name="delta" value="0.1"/>
    <param name="maxUrange" value="10.0"/>
    <param name="sigma" value="0.05"/>
    <param name="kernelSize" value="3"/>
    <param name="lstep" value="0.05"/>
    <param name="astep" value="0.05"/>
    <param name="iterations" value="5"/>
    <param name="lsigma" value="0.075"/>
    <param name="ogain" value="3.0"/>
    <param name="minimumScore" value="100"/>
    <param name="maximumScore" value="0"/>
    <param name="publish_map" value="true"/>	

  
  </node>

  </group>

    <!-- Import the fredbots package -->
  <!-- <arg name="atom_move_script" default="$(find fredbots)/src/scripts/atom_move.py" />
  <arg name="rviz_config" default="$(find fredbots)/src/atom/default.rviz" /> -->

  <!-- Run the atom_move.py node for each robot -->
  <!-- <node name="atom_move_robot1" pkg="fredbots" type="atom_1_move.py" output="screen" />
  <node name="atom_move_robot2" pkg="fredbots" type="atom_2_move.py" output="screen" />
  <node name="atom_move_robot3" pkg="fredbots" type="atom_3_move.py" output="screen" />
  <node name="atom_move_robot4" pkg="fredbots" type="atom_4_move.py" output="screen" /> -->

  <!-- Launch RViz with the provided configuration -->
  <arg name="rviz_config" default="$(find fredbots)/worlds/multi_robot.rviz" />
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rviz_config)" />

  <!-- launching the mapper node-->
  <!-- <node name="auto_map" pkg="fredbots" type="auto_map.py" output="screen" /> -->
  <node name="atom_move_robot4" pkg="fredbots" type="atom_4_move.py" output="screen" />

</launch>